keanu-project/src/main/java/io/improbable/keanu/tensor/TensorShapeValidation.java

```diff
@@ -77,18 +75,21 @@ public static void checkTensorsAreScalar(String message, long[]... shapes) {
      *
      * @param shapes the tensors for shape checking
      * @return either a scalar shape OR the single non-scalar shape.
-     * @throws IllegalArgumentException if there is more than one non-scalar shape
+     * @throws IllegalArgumentException if there is more than one non-scalar shape or multiple ranks of length 1 shapes
      */
-    public static long[] checkHasSingleNonScalarShapeOrAllScalar(long[]... shapes) {
-        Set<TensorShape> nonScalarShapes = getNonScalarShapes(shapes);
+    public static long[] checkHasOneNonSingularShapeOrAllSingular(long[]... shapes) {
+        Set<TensorShape> nonLengthOneShapes = getNonLengthOneShapes(shapes);
+        List<TensorShape> lengthOneShapes = getLengthOneShapes(shapes);
 
-        if (nonScalarShapes.isEmpty()) {
-            return Tensor.SCALAR_SHAPE;
-        } else if (nonScalarShapes.size() == 1) {
-            return nonScalarShapes.iterator().next().getShape();
-        } else {
-            throw new IllegalArgumentException("Shapes must match or be scalar");
+        if (nonLengthOneShapes.isEmpty()) {
+            if (!lengthOneShapes.isEmpty()) {
+                return lengthOneShapes.get(0).getShape();
```

Comment:
The length one shapes are sorted by rank because we have the situation where for example: 

a = [[[2]]] , which is a 1,1,1 
b = [3], which is a 1
c = a * b = [[[6]]], which is a 1,1,1

In this case you want the highest rank length one. This is due to the fact that we allow users to effectively broadcast this length one, which means we need the resulting broadcasting shape. More complex broadcasting isn't actually allowed in the vertices at the moment because they will be seen as incompatible shapes.

Reply:
ok - I think it would be worth adding a comment in the code that says this